### ISSUE TITLE
wheels for ARM64 Linux

### ISSUE THREAD

**Issue by dalberto** [openai/tiktoken#23]
When attempting to install `tiktoken` in a the `python:3.11-slim` Docker container running in Docker for Mac, I get an error about a missing rust compiler( `error: can't find Rust compiler...`). 

I'm not familiar with the Rust toolchain but I believe it should be possible to compile for `aarch64` by add the following:

`rustup target add aarch64-unknown-linux-gnu` in [pyproject.toml](https://github.com/openai/tiktoken/blob/156eff92d2dc88e108bdece47e584b05859af557/pyproject.toml#L17).

As a workaround, I'm just building from source, but it'd be super nice to have pre-built wheels.

My workaround:

```sh
export PATH="/root/.cargo/bin:$PATH"
apt-get update && apt-get install -y --no-install-recommends curl gcc &&  curl https://sh.rustup.rs -sSf | sh -s -- -y && apt-get install --reinstall libc6-dev -y
```

**Comment by hauntsaninja** [comment]
Thanks for the suggestion, but not planning on providing aarch64 wheels for now. In order to build aarch64 wheels in Github Actions, I'd need to use emulation, which is super slow and a pain to setup.

I'll leave this issue open for a bit, if this proves to be a popular request I might reconsider.

**Comment by sjrl** [comment]
We would definitely appreciate this as well for use in our docker images in the [Haystack](https://github.com/deepset-ai/haystack) library

**Comment by aliencaocao** [comment]
+1, need the wheel

**Comment by frankdilo** [comment]
+1, without a wheel our team can't easily install this library in the local dev environment, making it a pain to use.

**Comment by kwood** [comment]
+1 on having the wheel! Developing on an M1 mac w/ Docker becomes difficult.
My (somewhat painful) workaround is to use a two-stage build like this:
```Dockerfile
FROM python:3.11-slim-bullseye as builder

# Tiktoken requires Rust toolchain, so building it in a separate stage
RUN apt-get update && apt-get install -y gcc curl
RUN curl https://sh.rustup.rs -sSf | sh -s -- -y && apt-get install --reinstall libc6-dev -y
ENV PATH="/root/.cargo/bin:${PATH}"
RUN pip install --upgrade pip && pip install tiktoken

FROM python:3.11-slim-bullseye
... 
# Copy pre-built packages from builder stage
COPY --from=builder /usr/local/lib/python3.11/site-packages/ /usr/local/lib/python3.11/site-packages/
# Now install packages as normal — tiktoken will already be installed and skipped if present in requirements.txt
RUN pip install --upgrade pip && pip install -r requirements.txt
...
```

**Comment by aiastia** [comment]
> 

RUN pip install --upgrade pip && pip3 install tiktoken:
[2318](https://github.com/aiastia-dockerhub/Openaibot/actions/runs/4262903769/jobs/7418964777#step:6:2321)
#28 226.3       running build_rust
[2319](https://github.com/aiastia-dockerhub/Openaibot/actions/runs/4262903769/jobs/7418964777#step:6:2322)
#28 226.3       cargo rustc --lib --message-format=json-render-diagnostics --manifest-path Cargo.toml --release -v --features pyo3/extension-module -- --crate-type cdylib
[2320](https://github.com/aiastia-dockerhub/Openaibot/actions/runs/4262903769/jobs/7418964777#step:6:2323)
#28 226.3           Updating crates.io index
[2321](https://github.com/aiastia-dockerhub/Openaibot/actions/runs/4262903769/jobs/7418964777#step:6:2324)
#28 226.3       error: `cargo rustc --lib --message-format=json-render-diagnostics --manifest-path Cargo.toml --release -v --features pyo3/extension-module -- --crate-type cdylib` failed with code -9
[2322](https://github.com/aiastia-dockerhub/Openaibot/actions/runs/4262903769/jobs/7418964777#step:6:2325)
#28 226.3       [end of output]

**Comment by suanmiao** [comment]
+1 on needing the wheel, tiktoken is essential for many libraries and it avoids people from building docker image locally.

**Comment by hauntsaninja** [comment]
There's enough interest here that this is definitely something we should make happen. For various reasons, I've been short on time recently so can't commit to a timeline — but if someone opens a PR I will review and merge.

I copied over a QEMU setup I have from a different C project in https://github.com/openai/tiktoken/pull/54 ; unfortunately it looks like it didn't just work. Maybe it's obvious to someone here what's wrong?

**Comment by jongwook** [comment]
@hauntsaninja It looks like a SIGKILL since it says exit code -9; is it possible that the runner's system memory was not enough? Could we try with a [larger runner](https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners#adding-a-larger-runner-to-an-enterprise)?

**Comment by aiastia** [comment]
> 
Hello, I want to run tiktoken on arm64 server through docker, what should I do wrong now I tried https://github.com/openai/tiktoken/issues/23#issuecomment-1437463984

But  https://github.com/openai/tiktoken/issues/23#issuecomment-1444599282

**Comment by messense** [comment]
> I copied over the QEMU setup I have from a different C project I maintain in #54 ; unfortunately it looks like it didn't just work. Maybe it's obvious to someone here what's wrong?

I think it might just need to set a env var `CARGO_NET_GIT_FETCH_WITH_CLI="true"` because of https://github.com/rust-lang/cargo/issues/10583

**Comment by Peins** [comment]
Me too.

**Comment by hauntsaninja** [comment]
I've released 0.3.1 along with aarch64 wheels. Thanks @messense for the cargo environment variable fix!
